Add analytics framework

ekinanp opened the pull request:
Resolves https://github.com/puppetlabs/wash/issues/383

ekinanp committed:
Add EntrySchema#SetEntryType

This is needed by analytics.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add analytics package

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add analytics wrappers to the plugin package

The analytics wrappers are a useful way to centralize analytics
reporting.

This commit also updates the activity package to cache recorders instead
of journalFiles. The former is more generic, and accounts for the fact
that recorders must also record method invocations on top of logging
activity.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the API with the analytics changes

The /analytics/screenview endpoint will be used by the subcommands to
report their invocations.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update FUSE with the analytics changes

Currently, we collect analytics when autocompleting entries. It'd be
nice if there was an easy way to detect when that's happening. That work
is beyond the scope of this PR, however.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Collect analytics for subcommand invocations

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on analytics/client.go:
```diff
@@ -0,0 +1,318 @@
+		// tid => Tracking ID
+		//
+		// TODO: Set this to the main Wash GA account's tracking ID.
+		"tid": "UA-144659575-2",
```
Will do this once everything's approved.

ekinanp commented on analytics/client.go:
```diff
@@ -0,0 +1,318 @@
+		// av => Application Version
+		//
+		// TODO: Fill this in with a proper value
+		"av": "1.0",
```
This is a bit hard because the version's set in `cmd.version`. Wondering if this might be worth a separate version package. Maybe properly setting the version could be a separate issue?

MikaelSmith replied:
Reporting application version is pretty useful. A separate package that we update makes sense.

ekinanp replied:
Should that be a separate issue or should I tack it on here? Think that may need some changes to homebrew-puppet.

MikaelSmith replied:
I see it as part of this issue.

ekinanp commented:
It'll take about a day for the data to show up in custom reports, so should probably hold off on merging until there's something we can view. However, I am seeing the data in the Realtime monitor of GA.

Note that we'll also need to create a separate view to see the usage data reported here (and also turn off "filter analytics sent from spam bots").

ekinanp commented on fuse/dir.go:
```diff
@@ -34,7 +34,7 @@ func (d *dir) children(ctx context.Context) (map[string]plugin.Entry, error) {
 	// Cache List requests. FUSE often lists the contents then immediately calls find on individual entries.
 	if plugin.ListAction().IsSupportedOn(updatedEntry) {
-		return plugin.CachedList(ctx, updatedEntry.(plugin.Parent))
+		return plugin.List(ctx, updatedEntry.(plugin.Parent))
```
It'd be nice if there's an easy way to detect when we're auto-completing vs. invoking something. That way, we don't report analytics for auto-complete (unless we want to).

MikaelSmith replied:
I'm not concerned about reporting for auto-complete.

MikaelSmith commented on plugin/entrySchema.go:
```diff
@@ -172,6 +174,44 @@ func (s *EntrySchema) SetMetadataSchema(obj interface{}) *EntrySchema {
+// Note that entry type is optional. It should only be set when you want analytics to
+// collect data on that entry.
+func (s *EntrySchema) SetEntryType(entryType string) *EntrySchema {
+	// We could try to get this information from the type ID; however, it is difficult
```
I don't feel like this is useful enough for the added complexity. I don't agree that we need to leave out simple directories; it's easy enough to filter them out explicitly, or we can omit entries that don't have useful attributes/metadata and don't implement Read/Exec. I don't think capturing whether a plugin's types are used is something the plugin author should need to think about.

I'm also not sure the schema should be considered fragile. It's user-facing, so we shouldn't refactor it on a whim, and we won't be capturing long-running data.

ekinanp replied:
Open to alternative ways of getting the resource type, or omitting it altogether.

I do agree that knowing what resources are most popular doesn't really have any practical meaning aside from being something "cool to know." Plugin makes sense -- if AWS is popular, we should maintain AWS and maintain it well. That automatically includes its resources. So, I think I agree with you in that we should eliminate tracking the resource type altogether and just stick to method invocations (since the latter's simple enough to track).

MikaelSmith replied:
I'd be fine with that for now.

ekinanp replied:
Ok. I think it's still useful for the journal to record the method invocations, so will record that stuff via typeID instead of entryType.

MikaelSmith commented on analytics/config.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func readAnalyticsConfigFile(path string) (Config, error) {
+	_, err := os.Stat(path)
```
I don't think calling Stat just before reading gives us any new information. ioutil.ReadFile should return the same error if it doesn't exist. So we should just call ReadFile.

ekinanp replied:
Oh yes, ReadFile should be good enough. Will update.

MikaelSmith commented on api/server.go:
```diff
@@ -108,6 +115,7 @@ func StartAPI(registry *plugin.Registry, mountpoint string, socketPath string) (
 
 	r := mux.NewRouter()
 
+	r.Handle("/analytics/screenview", screenviewHandler).Methods(http.MethodPost)
```
I find it kind of annoying that this shows up in the activity journal:
```
Aug  6 14:37:56.528 API: POST /analytics/screenview
Aug  6 14:37:56.532 API: POST /analytics/screenview complete
```

ekinanp replied:
I don't think it's that bad. But we could ignore "/analytics" endpoint pings if it is annoying.

MikaelSmith commented on cmd/internal/server/core.go:
```diff
@@ -129,6 +146,14 @@ func (s *Server) shutdown() {
+	go func() {
+		s.analyticsClient.Flush()
+	}()
+	<-ticker.C
```
This waits a full 250ms, right? We should stop the ticker if the goroutine completes before then.

MikaelSmith commented on cmd/root.go:
```diff
@@ -33,6 +37,76 @@ func toRunE(main commandMain) runE {
+}
+
+// GA => Google Analytics
+func waitForGARegistration() {
```
This seems like it should wait for the GA flush to complete, or FlushDuration, whichever comes first. Not the full FlushDuration.

ekinanp replied:
Yeah, makes sense. I'll add that.

MikaelSmith commented on cmd/rootMain.go:
```diff
@@ -187,6 +187,7 @@ func rootMain(cmd *cobra.Command, args []string) exitCode {
 		cmdutil.ErrPrintf("Unable to start server: %v\n", err)
 		return exitCode{1}
 	}
+	registerInvocationToGA(cmd, socketpath)
```
Having these run over the socket seems weird, and results in an extra journal entry for starting the Wash shell.

ekinanp replied:
Ah yeah, wanted to maintain the symmetry with the other commands. I guess we could also say "Server start" => "Someone is using Wash", so that this stuff could move to the `internal/server` package.

ekinanp commented on cmd/internal/server/core.go:
```diff
@@ -129,6 +146,14 @@ func (s *Server) shutdown() {
+	ticker := time.NewTicker(analytics.FlushDuration)
+	defer ticker.Stop()
+	go func() {
+		s.analyticsClient.Flush()
```
Think it was to prevent Flush from blocking the server process from the rest of its cleanup (in case the network takes too long). I'll add a comment about that.

MikaelSmith replied:
Yeah, as a timer that makes sense.

ekinanp commented on analytics/config.go:
```diff
@@ -0,0 +1,96 @@
+// $HOME/.puppetlabs/wash/analytics.yaml.
+func GetConfig() (config Config, err error) {
+	// Check if analytics is disabled via the environment before
+	// returning
```
The defer enforces precedence -- disabling analytics via environment has greater precedence than everything else. Otherwise, if we check environment first, then unmarshalling could overwrite our disabled value and the additional logic to check that could be annoying.

Will add a comment about this if you think it is OK.

MikaelSmith replied:
Why not just return immediately if the environment variable is set? If it's set we don't need an ID and I don't see a reason to create the analytics file.

ekinanp replied:
That also makes sense. Will update.

ekinanp committed:
Don't wait for the full FlushDuration

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Send the initial start-up ping to GA in the server object

It's simpler to define "Server started" => "Someone is using Wash."
Defining usage this way means we send the start-up ping in the server
object.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Fix analytics config parsing

This commit simplifies the logic a bit and avoids writing an analytics
config file when analytics is disabled.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Revert "Add EntrySchema#SetEntryType"

This reverts commit e9513916300c396d2444edc074402af294728be6. We revert
the commit because the additional complexity of properly supporting an
entry type is not worth the additional information of knowing what entry
type invokes what method.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Remove the remaining bits of entry type tracking

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on activity/core.go:
```diff
@@ -107,6 +107,8 @@ func Warnf(ctx context.Context, msg string, a ...interface{}) {
 // It then records the invocation to the journal identified by the ID at `activity.JournalKey`
 // in the provided context.
+//
+// Note that entryType will typically be the type ID.
```
I want to keep entryType as-is here in case we do decide that reporting the entry type is useful.

ekinanp commented:
I've addressed all the feedback except for the version package. Working on that next.

MikaelSmith reviewed (commented):
Looks good, just waiting on adding version. .travis.yml and homebrew-puppet will need to be updated for the new path. Might make it worth adding a simple script to do release builds.

ekinanp committed:
Add version package

This avoids an import cycle between the analytics and cmd packages.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Set the tracking ID to the Wash account's tracking ID

Signed-off-by: Enis Inan <enis.inan@puppet.com>

MikaelSmith commented:
Oh, just realized we need to add a mention of this to the installation docs.

ekinanp commented:
Yeah, was about to do that.
